src/core/geometry/qgslinestring.cpp

```diff
@@ -2088,6 +2088,43 @@ bool QgsLineString::deleteVertex( QgsVertexId position )
   return true;
 }
 
+bool QgsLineString::deleteVertices( QList<QgsVertexId> positions )
+{
+  std::sort( positions.begin(), positions.end(), []( const QgsVertexId & a, const QgsVertexId & b )
+  {
+    return a.vertex > b.vertex;
+  }
+           );
+
+  for ( QgsVertexId position : positions )
+  {
+    if ( position.vertex >= mX.size() || position.vertex < 0 )
+    {
+      return false;
+    }
```

Comment:
Since positions is sorted, we can check `positions.constFirst().vertex >= mX.size()` and `positions.constLast().vertex < 0` once before starting the iteration.